tests/random/prng_test.go

```diff
@@ -59,116 +118,312 @@ func TestRead(t *testing.T) {
-					Counter: big.NewInt(int64(tt.bitsPerSample*tt.numSamples/256 + 1)),
-					Remain:  big.NewInt(256 - int64(tt.bitsPerSample*tt.numSamples)%256),
-				}
+						expectedSum := distMean * float64(tt.numSamples)
+						// The variance of as sum of uniform random variables in
+						// [0,sampleMax] is given by the sum of the individual variances
+						twoSigmaRelDev := 2 * math.Sqrt(distVar*float64(tt.numSamples)) / expectedSum
 
-				// PRNG increments the counter and then fills the entropy pool
-				// with keccak256(seed||counter).
-				entropy := new(big.Int).Lsh(wantState.Seed, 256)
-				entropy.Add(entropy, wantState.Counter)
-				// It's important not to use entropy.Bytes() here as we MUST
-				// have exactly 64 bytes to be hashed.
-				buf := make([]byte, 64)
-				entropy.FillBytes(buf)
-				// Some of the entropy has depleted; we know how much from
-				// Remain.
-				wantState.Entropy = new(big.Int).Rsh(
-					new(big.Int).SetBytes(crypto.Keccak256(buf)),
-					uint(256-wantState.Remain.Uint64()),
-				)
-
-				if diff := cmp.Diff(wantState, state, ethtest.Comparers()...); diff != "" {
-					t.Errorf("After %d samples of %d bits each; internal state diff (-want +got):\n%s", tt.numSamples, tt.bitsPerSample, diff)
-				}
-			})
+						if dev := 1 - float64(expectedSum)/float64(sum); math.Abs(dev) > twoSigmaRelDev {
+							t.Errorf("Sum = %s deviates by %.2f%% (>%.2f%% abs = 2 sigma) from expected sum %s", humanize.Comma(int64(sum)), dev*100, twoSigmaRelDev*100, humanize.Comma(int64(expectedSum)))
+						}
+					})
+
+					t.Run("Monobit Frequency", func(t *testing.T) {
+						expectedMonocount := float64(tt.numSamples*tt.bitsPerSample) / 2
+						// Follows a binomial distribution. Variance is nSamples/4
+						twoSigmaRelDev := 2. / math.Sqrt(float64(tt.numSamples*tt.bitsPerSample))
+
+						if dev := 1 - float64(expectedMonocount)/float64(onesCount); math.Abs(dev) > twoSigmaRelDev {
+							t.Errorf("One count = %s deviates by %.2f%% (>%.2f%% abs = 2 sigma) from expected sum %s", humanize.Comma(int64(onesCount)), dev*100, twoSigmaRelDev*100, humanize.Comma(int64(expectedMonocount)))
+						}
+					})
+
+					t.Run("Monobit CUSUM", func(t *testing.T) {
+						// See Sect. 2.13 in https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-22r1a.pdf
+						p := float64(1)
+						n := tt.numSamples * tt.bitsPerSample
+						z := cusumMax
+						sqrtn := math.Sqrt(float64(n))
+						kStart := int64(0.25 * (-float64(n)/float64(z) + 1))
+						kEnd := int64(0.25 * (float64(n)/float64(z) - 1))
+
+						for k := kStart; k <= kEnd; k++ {
+							p -= normalCDF(float64((4*k+1)*z)/sqrtn) - normalCDF(float64((4*k-1)*z)/sqrtn)
+						}
+
+						kStart = int64(0.25 * (-float64(n)/float64(z) - 3))
+						for k := kStart; k <= kEnd; k++ {
+							p += normalCDF(float64((4*k+3)*z)/sqrtn) - normalCDF(float64((4*k+1)*z)/sqrtn)
+						}
+
+						if p < 0.01 {
+							t.Errorf("Cusum test failed with p=%.2f < 0.01", p)
+						}
```

Comment:
I totally understand this without using the reference!

Reply:
Same!